// parcel-namer-rewrite/lib/Config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Config = void 0;

var _logger = require("@parcel/logger");

var _path = _interopRequireDefault(require("path"));

var _fs = _interopRequireDefault(require("fs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const PACKAGE_JSON_SECTION = "parcel-namer-rewrite";

class Config {
  /**
   * Disable in development
   */

  /**
   * Name hashing
   */
  // "never", "always"

  /**
   * Disable logging names
   */

  /**
   * Use file name hashes from parcel
   */
  constructor() {
    _defineProperty(this, "profileEnvKey", 'PARCEL-NAMER-REWRITE-PROFILE');

    _defineProperty(this, "rules", void 0);

    _defineProperty(this, "chain", void 0);

    _defineProperty(this, "disable", false);

    _defineProperty(this, "hashing", '');

    _defineProperty(this, "silent", false);

    _defineProperty(this, "useParcelHash", true);

    this.chain = '@parcel/namer-default';
    this.rules = [];
  }

  loadFromPackageFolder(rootFolder, env, defaultProfiles, logger) {
    const packageJson = _fs.default.readFileSync(_path.default.join(rootFolder, 'package.json')).toString();

    const packageInfo = JSON.parse(packageJson);
    const packageSection = packageInfo[PACKAGE_JSON_SECTION];

    if (!packageSection) {
      logger.warn({
        message: `no "${PACKAGE_JSON_SECTION}" section in package.json. Use no-rules config`
      });
      return;
    }

    this._loadFromPackageSection(packageSection, logger);

    let profileNames = '';
    if (defaultProfiles) profileNames += ',' + defaultProfiles;
    if (this.profileEnvKey in env) profileNames += ',' + env[this.profileEnvKey];
    const profileSections = packageSection['profiles'];

    if (profileSections) {
      profileNames.split(/[,]/).map(profile => profileSections[profile]).filter(profile => !!profile).forEach(profile => {
        this._loadFromPackageSection(profile, logger);
      });
    }
  }

  _loadFromPackageSection(section, logger) {
    if (!section) return;

    if ('profile-env-key' in section) {
      this.profileEnvKey = section['profile-env-key'];
    }

    if ('chain' in section) {
      this.chain = section.chain;
    }

    if ('neverHashing' in section) {
      this.neverHashing = !!section.neverHashing;
    }

    this.silent = 'silent' in section && section.silent;

    if ('useParcelHash' in section) {
      this.useParcelHash = !!section.useParcelHash;
    }

    if ('rules' in section) {
      Object.keys(section.rules).forEach(k => {
        const ruleData = section.rules[k];
        const ruleTo = typeof ruleData === 'string' ? ruleData : null;

        if (ruleTo === null) {
          logger.warn(`No "to" rule for test "${k}" `);
          return;
        }

        this.rules.push({
          test: new RegExp(k),
          to: ruleTo
        });
      });
    }

    if ('developmentHashing' in section) {
      throw Error(`The "developmentHashing" option is not supported any more. Add "development" profile and set {"hashing": "never"} there. See documentation for details here: https://github.com/ol-loginov/parcel-namer-rewrite`);
    }

    if ('developmentDisable' in section) {
      throw Error(`The "developmentDisable" option is not supported any more. Add "development" profile and set {"disable": false} there. See documentation for details here: https://github.com/ol-loginov/parcel-namer-rewrite`);
    }

    if ('hashing' in section) {
      this.hashing = section['hashing'];
    }

    if ('disable' in section) {
      this.disable = !!section['disable'];
    }
  }

  selectRule(name) {
    const matches = this.rules.map(rule => rule.test.test(name) ? rule : null).filter(rule => rule != null);

    if (matches.length > 0) {
      return matches[0];
    }

    return null;
  }

}

exports.Config = Config;
